DataStructures/Lists/CircularBuffer.cs

```diff
@@ -0,0 +1,197 @@
+        /// Initializes a circular buffer with given length
+        /// </summary>
+        /// <param name="length">The length of the buffer</param>
+        public CircularBuffer(int length, bool canOverride=true) 
+        {
+            if (length < 1) 
+            {
+                throw new ArgumentOutOfRangeException("length can not be zero or negative");
+            }
+            _circularBuffer = new T[length + 1];
+            _default = _circularBuffer[length];
+            _end = 0;
+            _start = 0;
+            CanOverride = canOverride;
+        }
+
+        /// <summary>
+        /// Writes value to the back of the buffer
+        /// </summary>
+        /// <param name="value">value to be added to the buffer</param>
+        public void Add(T value) 
+        {
+            if (CanOverride) 
+            {
+                innerInsert(value);
+            }
+            else 
+            {
+                if (IsFilledUp) 
+                {
+                    throw new CircularBufferFullException($"Circular Buffer is filled up. {value} can not be inserted");
+                }
+                innerInsert(value);
+            }
+        }
+
+        // Inserts data into the buffer without checking if it is full
+        private void innerInsert(T value) 
+        {
+            _circularBuffer[_end] = value;
+            _end = (_end + 1) % _circularBuffer.Length;
+            if (_end == _start) 
+            {
+                _start = (_start + 1) % _circularBuffer.Length;
+            }
+        }
+
+        /// <summary>
+        ///     Reads and removes the value in front of the buffer, and places the next value in front.
+        /// </summary>
+        public T Pop() 
+        {
+            var result = _circularBuffer[_start];
+            _circularBuffer[_start] = _circularBuffer[_end];
+            _start = (_start + 1) % _circularBuffer.Length;
+            return result;
+        }
+
+        #region IEnumerable Implementation
+        public IEnumerator<T> GetEnumerator() 
+        {
+            for (int i = 0; i < Count; i++) {
+                yield return _circularBuffer[i];
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() 
+        {
+            return GetEnumerator();
+        }
+        #endregion
+
+        #region ICollection Implementation
+        /// <summary>
+        /// Returns the number of elements.
+        /// </summary>
+        public int Count 
+        {
+            get { return _length; }
+        }
+        /// <summary>
+        /// Checks whether this collection is readonly
+        /// </summary>
+        public bool IsReadOnly 
+        {
+            get { return false; }
+        }
+        /// <summary>
+        /// Clears this instance
+        /// </summary>
+        public void Clear() 
+        {
+            for (int i = 0; i < _circularBuffer.Length; i++) 
+            {
+                _circularBuffer[i] = _default;
```

Comment:
Why a for loop and not create a new one?
```
_circularBuffer = new T[length + 1];
```

Reply:
Implemented